From: Luca Weiss <luca.weiss@fairphone.com>
Date: Mon, 18 Nov 2024 20:45:31 +0100
Subject: arm64: dts: qcom: qcm6490-fairphone-fp5: Add UW rear camera support

Add the necessary nodes to enable the ultra-wide rear camera (Sony
IMX858).

Signed-off-by: Luca Weiss <luca.weiss@fairphone.com>
---
 arch/arm64/boot/dts/qcom/qcm6490-fairphone-fp5.dts | 47 +++++++++++++++++++++-
 1 file changed, 46 insertions(+), 1 deletion(-)

diff --git a/arch/arm64/boot/dts/qcom/qcm6490-fairphone-fp5.dts b/arch/arm64/boot/dts/qcom/qcm6490-fairphone-fp5.dts
index e1f32bd..b633a66 100644
--- a/arch/arm64/boot/dts/qcom/qcm6490-fairphone-fp5.dts
+++ b/arch/arm64/boot/dts/qcom/qcm6490-fairphone-fp5.dts
@@ -616,6 +616,15 @@ &camss {
 	status = "okay";
 
 	ports {
+		port@1 {
+			csiphy1_ep: endpoint {
+				clock-lanes = <7>;
+				data-lanes = <0 1 2 3>;
+				bus-type = <MEDIA_BUS_TYPE_CSI2_DPHY>;
+				remote-endpoint = <&camera_imx858_ep>;
+			};
+		};
+
 		port@3 {
 			csiphy3_ep: endpoint {
 				clock-lanes = <7>;
@@ -643,7 +652,36 @@ eeprom@50 {
 };
 
 &cci0_i2c1 {
-	/* IMX858 @ 29 */
+	/* actuator@e (dw9800k) */
+
+	camera@29 {
+		compatible = "sony,imx858";
+		reg = <0x29>;
+
+		vdig-supply = <&vreg_l1p>;
+		vana-supply = <&vreg_l4p>;
+		vif-supply = <&vreg_l6p>;
+
+		clocks = <&camcc CAM_CC_MCLK1_CLK>;
+		assigned-clocks = <&camcc CAM_CC_MCLK1_CLK>;
+		assigned-clock-rates = <24000000>;
+
+		reset-gpios = <&tlmm 21 GPIO_ACTIVE_LOW>;
+		pinctrl-0 = <&cam_mclk1_default>;
+		pinctrl-names = "default";
+
+		orientation = <1>; /* Rear facing */
+		rotation = <270>;
+
+		port {
+			camera_imx858_ep: endpoint {
+				data-lanes = <1 2 3 4>;
+				bus-type = <MEDIA_BUS_TYPE_CSI2_DPHY>;
+				link-frequencies = /bits/ 64 <600000000>;
+				remote-endpoint = <&csiphy1_ep>;
+			};
+		};
+	};
 
 	eeprom@54 {
 		compatible = "giantec,gt24p128f", "atmel,24c128";
@@ -1270,6 +1308,13 @@ bluetooth_enable_default: bluetooth-enable-default-state {
 		bias-disable;
 	};
 
+	cam_mclk1_default: cam-mclk1-default-state {
+		pins = "gpio65";
+		function = "cam_mclk";
+		drive-strength = <2>;
+		bias-disable;
+	};
+
 	cam_mclk3_default: cam-mclk3-default-state {
 		pins = "gpio67";
 		function = "cam_mclk";
